source/base/polynomials_bernardi_raugel.cc

```diff
@@ -0,0 +1,254 @@
+                                  bubble_grad_grads,
+                                  bubble_third_derivatives,
+                                  bubble_fourth_derivatives);
+  polynomial_space_Q.compute(unit_point,
+                             Q_values,
+                             Q_grads,
+                             Q_grad_grads,
+                             Q_third_derivatives,
+                             Q_fourth_derivatives);
+
+  // first dim*vertices_per_cell functions are Q_1^2 functions
+  for (unsigned int i = 0; i < dim * GeometryInfo<dim>::vertices_per_cell; ++i)
+    {
+      if (values.size() != 0)
+        {
+          values[i] = units[i % dim] * Q_values[i / dim];
+        }
+      if (grads.size() != 0)
+        {
+          grads[i] = outer_product(units[i % dim], Q_grads[i / dim]);
+        }
+      if (grad_grads.size() != 0)
+        {
+          grad_grads[i] = outer_product(units[i % dim], Q_grad_grads[i / dim]);
+        }
+      if (third_derivatives.size() != 0)
+        {
+          third_derivatives[i] =
+            outer_product(units[i % dim], Q_third_derivatives[i / dim]);
+        }
+      if (fourth_derivatives.size() != 0)
+        {
+          fourth_derivatives[i] =
+            outer_product(units[i % dim], Q_fourth_derivatives[i / dim]);
+        }
+    }
+
+  // last faces_per_cell functions are bubble functions
+  for (unsigned int i = dim * GeometryInfo<dim>::vertices_per_cell;
+       i < dim * GeometryInfo<dim>::vertices_per_cell +
+             GeometryInfo<dim>::faces_per_cell;
+       ++i)
+    {
+      unsigned int j =
+        i -
+        dim *
+          GeometryInfo<dim>::vertices_per_cell; // ranges 0 to faces_per_cell-1
+      if (values.size() != 0)
+        {
+          values[i] = normals[j] * bubble_values[aniso_indices[j]];
+        }
+      if (grads.size() != 0)
+        {
+          grads[i] = outer_product(normals[j], bubble_grads[aniso_indices[j]]);
+        }
+      if (grad_grads.size() != 0)
+        {
+          grad_grads[i] =
+            outer_product(normals[j], bubble_grad_grads[aniso_indices[j]]);
+        }
+      if (third_derivatives.size() != 0)
+        {
+          third_derivatives[i] =
+            outer_product(normals[j],
+                          bubble_third_derivatives[aniso_indices[j]]);
+        }
+      if (fourth_derivatives.size() != 0)
+        {
+          fourth_derivatives[i] =
+            outer_product(normals[j],
+                          bubble_fourth_derivatives[aniso_indices[j]]);
+        }
+    }
+}
+
+template <int dim>
+unsigned int
+PolynomialsBernardiRaugel<dim>::compute_n_pols(unsigned int k)
+{
+  (void)k;
+  Assert(k == 1, ExcNotImplemented());
+  if (dim == 2 || dim == 3)
+    return dim * GeometryInfo<dim>::vertices_per_cell +
+           GeometryInfo<dim>::faces_per_cell;
+  // 2*4+4=12 polynomials in 2D and 3*8+6=30 polynomials in 3D
+
+  Assert(false, ExcNotImplemented());
```

Comment:
Would you mind putting this in the constructor too?

Reply:
Since this function is called by the second member in the member initializer list in the constructor, I'm hesitant because feel it would be a little redundant since it would never reach the statement in the constructor. Unless it's so that users can see it upfront :smiley:

Reply:
I reread the constructor again and the same assertions are already there. I am not sure how I missed them the first time.

Reply:
No worries!